text/0000-infer-outlives.md

```diff
@@ -0,0 +1,578 @@
+
+```rust
+struct Foo<'b, U> {
+  bar: Bar<'b, U>
+}
+
+struct Bar<'a, T> where T: 'a {
+  x: &'a (),
+  y: T
+}
+```
+
+Here, `Bar` is declared with the where clause that `T: 'a`. This
+results in the requirement that `(T: 'a) in A[Bar]`. `Foo`, meanwhile,
+requires that any outlives requirements for `Bar<'b, U>` are
+satisfied, and hence as the rule that `('a => 'b, T => U) (A[Bar]) <=
+A[Foo]`. The minimal solution to this is:
+
+- `A[Foo] = (U: 'b)`
+- `A[Bar] = (T: 'a)`
+
+This means that we would infer an implicit outlives requirements of
+`U: 'b` for `Foo`; for `Bar` we would infer `T: 'a` but that was
+explicitly declared.
+
+### Example 4: Normalization or lack thereof
+
+Let us revisit the case where the where-clause is due
+to an impl:
+
+```rust
+trait MakeRef<'a> {
+  type Type;
+}
+
+impl<'a, T> MakeRef<'a> for Vec<T>
+  where T: 'a
+{
+  type Type = &'a T;
+}
+
+struct Foo<'a, T> { // Results in an error
+  foo: <Vec<T> as MakeRef<'a>::Type
+}
+```
+
+Here, for the struct `Foo<'a, T>`, we will in fact create no
+constraints for its where-clause set, and hence we will infer an empty
+set. This is because we encounter the field type `<Vec<T> as
+MakeRef<'a>>::Type`, and in such a case we ignore the trait reference
+itself and just require that `Vec<T>` is well-formed, which does not
+result in any outlives requirements as it contains no references. 
+
+Now, when we go to check the full well-formedness rules for `Foo`, we will
+get an error -- this is because, in that context, we will try to normalize
+the associated type reference, but we will fail in doing so because we do not
+have any where-clause stating that `T: 'a` (which the impl requires).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary drawbacks were covered in depth in the guide-level explanation,
+which also covers why they are not considered to be major problems:
+
+- Long-range errors
+  - can be readily mitigated by better explanations
+- Removing fields can affect semver compatibility
+  - considered unlikely to occur frequently in practice
+  - already true that changing field types can affect semver compatibility
+  - semver-like tool could help to mitigate
+
+# Rationale and Alternatives
+[alternatives]: #alternatives
+
+Naturally, we might choose to retain the status quo, and continue to
+require outlives annotations on structs. Assuming however that we wish
+to remove them, the primary alternative is to consider going *farther*
+than this RFC in various ways.
+
+We might make try to infer outlives requirements for impls as well,
+and thus eliminate the final place where `T: 'a` requirements are
+needed. However, this would introduce complications in the
+implementation -- in order to propagate requirements from impls to
+structs, we must be able to do associated type normalization and hence
+trait solving, but we would have to do before we know the full WF
+requirements for each struct. The current setup avoids this
+complication.
```

Comment:
I think "might make try to" has one verb too many and "we would have to do before we know" is missing a verb.